File: Leetcode/849-Maximize-Distance-to-Closest-Person/MaximizeDistancetoClosestPerson.cpp

class Solution {
public:
    int maxDistToClosest(vector<int>& seats) {
        int max = maxZeroDistance(seats);
        if (max % 2 == 1) {
            return max/2 + 1;
        }
        return max/2;
    }
private:
    int maxZeroDistance(vector<int>& seats) {
        bool isBeginning = true;
        int max = 0;
        int count = 0;
        for(int i=0; i<seats.size(); i++) {
            if (seats[i] == 0) {
                count++;
            } else {
                if (isBeginning) {
                    max = std::max(max, count*2);
                } else {
                    max = std::max(max, count);
                }
                isBeginning = false;
                count = 0;
            }
        }
        if (count != 0) {
            max = std::max(max, count*2);
        }
        return max;
    }
};

File: Leetcode/849-Maximize-Distance-to-Closest-Person/MaximizeDistancetoClosestPerson.go
func maxDistToClosest(seats []int) int {
	res := maxZeroDistance(seats)
	if res%2 == 1 {
		return res/2 + 1
	}
	return res / 2
}

func maxZeroDistance(seats []int) int {
	isB := true
	res := 0
	count := 0
	for i := 0; i < len(seats); i++ {
		if seats[i] == 0 {
			count++
		} else {
			if isB {
				res = max(res, count*2)
			} else {
				res = max(res, count)
			}
			isB = false
			count = 0
		}
	}
	if count != 0 {
		res = max(res, count*2)
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
